core/src/main/java/hudson/model/AbstractItem.java

```diff
@@ -577,9 +585,102 @@ public void delete( StaplerRequest req, StaplerResponse rsp ) throws IOException
      */
     public void delete() throws IOException, InterruptedException {
         checkPermission(DELETE);
-        synchronized (this) { // could just make performDelete synchronized but overriders might not honor that
-            performDelete();
-        } // JENKINS-19446: leave synch block, but JENKINS-22001: still notify synchronously
+        boolean responsibleForAbortingBuilds = !ItemDeletion.contains(this);
+        boolean ownsRegistration = ItemDeletion.register(this);
+        if (!ownsRegistration && ItemDeletion.isRegistered(this)) {
+            // we are not the owning thread and somebody else is concurrently deleting this exact item
+            throw new Failure(Messages.AbstractItem_BeingDeleted(getPronoun()));
```

Comment:
Maybe it worth putting this warning into UI as well (e.g. in the summary page near the "disabled" status)